lib/rules/string-quotes/index.js

```diff
@@ -23,31 +28,117 @@ const rule = function(expectation) {
+          break;
+        case "decl":
+          checkDeclOrAtRule(node, node.value, declarationValueIndex);
+          break;
+        case "rule":
+          checkRule(node);
+          break;
+      }
+    });
 
-    // allow @charset rules to have double quotes, in spite of the configuration
-    if (erroneousQuote === '"') {
-      root.walkAtRules(atRule => {
-        if (atRule.name === "charset") {
-          const badAtRule = atRule.toString();
-          const goodAtRule = badAtRule.split('"').join("'");
-          cssString = cssString.replace(badAtRule, goodAtRule);
-        }
+    function checkRule(rule) {
+      if (!isStandardSyntaxRule(rule)) {
+        return;
+      }
+      if (
+        rule.selector.indexOf("[") === -1 ||
+        rule.selector.indexOf("=") === -1
+      ) {
+        return;
+      }
+
+      const fixPositions = [];
+      parseSelector(rule.selector, result, rule, selectorTree => {
+        selectorTree.walkAttributes(attributeNode => {
+          if (
+            attributeNode.quoted &&
+            attributeNode.value.indexOf(erroneousQuote) !== -1
+          ) {
+            const openIndex =
+              // index of the start of our attribute node in our source
+              attributeNode.sourceIndex +
+              // length of our attribute
+              attributeNode.attribute.length +
+              // length of our operator , ie '='
+              attributeNode.operator.length +
+              // and the length of the quote
+              erroneousQuote.length;
+
+            if (context.fix) {
+              const closeIndex =
+                // our initial index
+                openIndex +
+                // the length of our value
+                attributeNode.value.length -
+                // with the length of our quote subtracted
+                erroneousQuote.length;
+              fixPositions.push(openIndex, closeIndex);
+            } else {
+              report({
+                message: messages.expected(expectation),
+                node: rule,
+                index: openIndex,
+                result,
+                ruleName
+              });
+            }
+          }
+        });
+      });
+      fixPositions.forEach(fixIndex => {
+        rule.selector = replaceQuote(rule.selector, fixIndex, correctQuote);
       });
     }
 
-    styleSearch({ source: cssString, target: erroneousQuote }, match => {
-      report({
-        message: messages.expected(expectation),
-        node: root,
-        index: match.startIndex,
-        result,
-        ruleName
+    function checkDeclOrAtRule(node, value, getIndex) {
+      const fixPositions = [];
+      // Get out quickly if there are no erroneous quotes
+      if (value.indexOf(erroneousQuote) === -1) {
+        return;
+      } else if (node.type === "atrule" && node.name === "charset") {
+        // allow @charset rules to have double quotes, in spite of the configuration
+        return;
```

Comment:
There is an issue related this specific condition https://github.com/stylelint/stylelint/issues/2788

Maybe we should report it if `@charset` has single quotes, and replace them to double quotes with `--fix`? Or it should be a separate PR?

Reply:
i think a separate pr would make sense - this is a pretty specific case. i just tried to keep the original logic in my pr.